packages/zoe/src/contracts/loan/close.js

```diff
@@ -0,0 +1,87 @@
+// @ts-check
+
+import './types';
+
+import { assert, details } from '@agoric/assert';
+
+import { assertProposalShape, trade } from '../../contractSupport';
+
+// The debt, the amount which must be repaid, is just the amount
+// loaned plus interest (aka stability fee). All debt must be repaid
+// in one offer. In exchange for total repayment, all collateral is
+// given back and the contract is shutdown.
+
+/** @type {MakeCloseLoanInvitation} */
+export const makeCloseLoanInvitation = (zcf, config) => {
+  const { collateralSeat, getDebt, lenderSeat } = config;
+
+  /** @type {OfferHandler} */
+  const repayAndClose = repaySeat => {
+    assertProposalShape(repaySeat, {
+      give: { Loan: null },
+      want: { Collateral: null },
+    });
+
+    const loanMath = zcf.getTerms().maths.Loan;
+    const loanBrand = zcf.getTerms().brands.Loan;
+    const collateralBrand = zcf.getTerms().brands.Collateral;
+
+    const repaid = repaySeat.getAmountAllocated('Loan', loanBrand);
+
+    // This must be a function because the amount of debt will change
+    // over time as interest is added.
+    const debt = getDebt();
+
+    // All debt must be repaid.
+    assert(
+      loanMath.isGTE(repaid, debt),
+      details`Not enough Loan assets have been repaid.  ${debt} is required, but only ${repaid} was repaid.`,
+    );
+
+    // We cannot use `swap` or `swapExact` helper because the collateralSeat
+    // doesn't have a `want`.
+
+    // Transfer the collateral to the repaySeat and remove the
+    // required Loan tokens. Any excess Loan tokens are kept by the repaySeat.
+    trade(
+      zcf,
+      {
+        seat: repaySeat,
+        gains: {
+          Collateral: collateralSeat.getAmountAllocated(
+            'Collateral',
+            collateralBrand,
+          ),
+        },
+      },
+      {
+        seat: collateralSeat,
+        gains: { Loan: debt },
+      },
+    );
+
+    // Transfer the repaid loan tokens to the lender
+    trade(
+      zcf,
+      {
+        seat: lenderSeat,
+        gains: { Loan: debt },
+      },
+      {
+        seat: collateralSeat,
+        gains: {},
+      },
+    );
```

Comment:
Can we do these transfers in a single reallocate? Does anyone have a `want` that gets in the way?

Or is it simpler with two trades because we don't have to do any of the subtraction? 

Did you compare the two approaches?

Reply:
I think it's simpler with two trades. There is no `want` that gets in the way, because the `lenderSeat` and the `collateralSeat` do not have a `want`. I can try out the single reallocate to check and see if it's actually simpler.